src/main/java/org/apache/commons/io/input/AutoCloseInputStream.java

```diff
@@ -66,19 +68,47 @@ public void close() throws IOException {
     }
 
     /**
-     * Automatically closes the stream if the end of stream was reached.
+     * Automatically closes the stream if the end of stream was reached unless the stream was marked.
      *
      * @param n number of bytes read, or -1 if no more bytes are available
      * @throws IOException if the stream could not be closed
      * @since 2.0
      */
     @Override
     protected void afterRead(final int n) throws IOException {
-        if (n == EOF) {
+        if (n == EOF && !marked) {
```

Comment:
This changes the contract of the method and the whole class. The class Javadoc states "Proxy stream that closes and discards the underlying stream as soon as the end of input has been reached or when the stream is explicitly closed." If feels like you need a different class. Now, it is possible that the original author did not even think of the mark and reset use cases, but still, this has the potential of defeating the purpose of the class: which is to auto close no matter what. Maybe you need a class called "AutoCloseUnmarkedInputStream"?

Reply:
The thing is AutoCloseInputStream itself does not respect the contract of mark/rest API (which makes Tika confuse) and I feel this use case was simply forgotten.

IMO either AutoCloseInputStream support mark/rest or it should force override mark/reset and return false for markSupported().

Reply:
I do see your point and let me rephrase it to make sure I understand: Either we should implement the mark/reset APIs properly or we should implement markSupported() to return false. 

My concern remains for this use case, today: 
- I open a stream
- I read
- I call mark() and read() here and there
- I read to the end of stream and the stream closes as it should
- I do not bother to call close()
- All is well

With this change, my random calls to mark() before I read to the end of the stream cause the stream to remain open. Clearly a leak. Granted, I _should_ close a stream that I open somewhere or in a try-with-resources block but this class lives outside of this use case.

It seems to me like the simplest solution would be for markSupported() to return false. WDYT?

Reply:
My main point is that it cannot stay like it currently is.

So yes if you want to be sure mark() can be a random mistake then mark/reset should throw an exception and markSupported() should always return false.

I can change my pull request in that directly, no problem. Both works.